hs-bindgen/src-internal/HsBindgen/Artefact.hs

```diff
@@ -103,21 +132,24 @@ data ArtefactEnv = ArtefactEnv {
 --
 -- All top-level artefacts will be cached (this is not true for computed
 -- artefacts, using, for example, the 'Functor' interface, or 'Lift').
+--
 runArtefacts :: forall e a.
      Tracer ArtefactMsg
   -> IORef (TracerState e)
   -> BootArtefact
   -> FrontendArtefact
   -> BackendArtefact
   -> Artefact a
-  -> IO (Either (TraceException e) a)
+  -> IO (Either (TraceException e) a, [FileSystemAction])
 runArtefacts
   tracer
   tracerStateRef
   BootArtefact{..}
   FrontendArtefact{..}
   BackendArtefact{..}
-  artefact = runReaderT (runExceptT (runArtefact artefact)) env
+  artefact = do
+    (result, _, actions) <- runRWST (runExceptT (runArtefact artefact)) env ()
+    return (result, actions)
```

Comment:
Is there a reason why you return the actions and not perform them here?

I have the feeling they should be performed here. This also makes sense in that the comment at `FileContent` describes what the file operations do, but the code is not in the module but in `HsBindgen.hs`.

Also, I mentioned this already but I think we didn't discuss it properly yet: Why do we use an exception instead of an error trace? To my knowledge, we should avoid panics if possible and think about `hs-bindgen` as if it were a compiler.

The `Bind` operation of `Artefact` checks for error traces, and we could use that machinery and just emit an error trace.